swarm/swap/protocol.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/protocols"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/log"
+)
+
+const (
+	IsActiveProtocol = true
+)
+
+//Here we define the Swap p2p.protocol
+//We use it to standardize the interaction between nodes who need clear their accounts.
+//Accounting itself is separate (see swarm/swap/swap.go)
+//This protocol focuses on sending and receiving cheques and
+//cheque management/clearing
+//For a better understanding, please read the Swap network paper "Generalized Swap swear and swindle games"
+type Protocol struct {
+	peersMu sync.RWMutex
+	peers   map[discover.NodeID]*Peer
+}
+
+//This is the peer representing a participant in this protocol
+type Peer struct {
+	*protocols.Peer
+}
+
+//Create a new protocol instance
+func NewProtocol() *Protocol {
+	proto := &Protocol{
+		peers: make(map[discover.NodeID]*Peer),
+	}
+	return proto
+}
+
+// NewPeer is the constructor for the protocol Peer
+func NewPeer(peer *protocols.Peer) *Peer {
+	p := &Peer{
+		Peer: peer,
+	}
+	return p
+}
+
+//In a peer exchange, if node A gets too indebted with node B,
+//node A issues a cheque and sends it to B
+type IssueChequeMsg struct {
+	Cheque *Cheque
+}
+
+//In a scenario where B received a cheque from A, node B can
+//redeem a cheque, which means it kicks off the process to cash it in.
+//In this case, this message is sent to peer A
+type RedeemChequeMsg struct {
```

Comment:
what is this about? you redeem through the blockhain

Reply:
I was looking at the sw3 paper "https://www.overleaf.com/project/59c298da5b3a6e3b7ec80d56", Figure 1, there redeeming seems to be initiated by the `B client`, so I assumed there would be some message exchange involved too.

Nevertheless, thinking of it better, it probably means that `B client` interacts with the blockchain directly, so this message may be obsolete indeed - although I am wondering if we should introduce this message nonetheless for better control of what's going on over the protocol

Reply:
just introduce stuff when needed